axoupdater/src/test/helpers.rs

```diff
@@ -0,0 +1,114 @@
+        .replace("BINARIES", &binaries)
+        .replace("PACKAGE", package)
+        .replace("APP_NAME", app_name)
+        .replace("INSTALL_PREFIX", &prefix.replace('\\', "\\\\"))
+        .replace("VERSION", version)
+}
+
+/// Generates an install receipt given the specified fields and writes it to disk at the appropriate location in `config_path`. The path to the new file is returned.
+pub fn write_receipt(
+    app_name: &str,
+    package: &str,
+    binaries: &[&str],
+    version: &str,
+    prefix: &PathBuf,
+    config_path: &PathBuf,
+) -> std::io::Result<PathBuf> {
+    let contents = install_receipt(
+        app_name,
+        package,
+        binaries,
+        version,
+        &prefix.to_string_lossy(),
+    );
+    let receipt_name = config_path.join(format!("{package}-receipt.json"));
+    std::fs::create_dir_all(&config_path)?;
+    std::fs::write(&receipt_name, contents)?;
+
+    Ok(receipt_name)
+}
+
+/// Performs an end-to-end run test for `app_name` within `package`. `bin` is the path to the executable being tested, usually the one from the `CARGO_BIN_EXE_<name>` environment variable.
+/// Rather than returning a value, this function performs assertions.
+/// Because it writes to real files outside a temporary directory, it's highly recommended that this only becalled within CI builds.
+pub fn perform_runtest(
+    app_name: &str,
+    package: &str,
+    bin: &PathBuf,
+    binaries: &[&str],
+    args: &[&str],
+) {
+    let home = homedir::get_my_home().unwrap().unwrap();
+
+    let app_home = &home.join(".cargo").join("bin");
+    let app_path = &app_home.join(format!("{app_name}{EXE_SUFFIX}"));
+
+    #[cfg(target_family = "unix")]
+    let config_path = &home.join(".config").join(app_name);
+    #[cfg(target_family = "windows")]
+    let config_path = &std::env::var("LOCALAPPDATA")
+        .map(std::path::PathBuf::from)
+        .unwrap()
+        .join(app_name);
+
+    // Ensure we delete any previous copy that may exist
+    // at this path before we copy in our version.
+    if app_path.exists() {
+        std::fs::remove_file(app_path).unwrap();
+    }
+    assert!(!app_path.exists());
+
+    // Install to the home directory
+    std::fs::copy(bin, app_path).unwrap();
+
+    // Create a fake install receipt
+    // We lie about being a very old version so we always
+    // consider upgrading to something.
+    write_receipt(
+        app_name,
+        package,
+        binaries,
+        "0.0.1",
+        &app_home,
+        &config_path,
+    )
+    .unwrap();
+
+    let output = Command::new(app_path)
+        .args(args)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .output()
+        .unwrap();
+
+    let out_str = String::from_utf8_lossy(&output.stdout);
+    let err_str = String::from_utf8_lossy(&output.stderr);
+
+    assert!(
+        output.status.success(),
+        "status code: {}, stdout: {out_str}; stderr: {err_str}",
+        output.status
+    );
+}
```

Comment:
Since this isn't going to a tempdir, can we return the path to let them do followup tests..? (check `--version` or whatever)

Reply:
Yeah, good call. Done.